processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -349,10 +382,53 @@ else if ( !method.isUpdateMethod() ) {
                 beforeMappingMethods,
                 afterMappingMethods,
                 finalizeMethod,
-                mappingReferences
+                mappingReferences,
+                subClasses
             );
         }
 
+        private SubClassMapping createSubClassMapping(SubClassMappingOptions subClassMappingOptions) {
+            TypeFactory typeFactory = ctx.getTypeFactory();
+            Type sourceType = typeFactory.getType( subClassMappingOptions.getSourceClass() );
+            Type targetType = typeFactory.getType( subClassMappingOptions.getTargetClass() );
+            SubClassMapping subClassMapping =
+                new SubClassMapping( sourceType, targetType, ctx.getTypeUtils() );
+
+            SourceRHS rightHandSide = new SourceRHS(
+                "subClassMapping",
+                sourceType,
+                Collections.emptySet(),
+                "SubClassMapping for " + sourceType.getFullyQualifiedName() );
+            SelectionCriteria criteria =
+                SelectionCriteria
+                                 .forMappingMethods(
+                                     new SelectionParameters(
+                                         Collections.emptyList(),
+                                         Collections.emptyList(),
+                                         subClassMappingOptions.getTargetClass(),
+                                         ctx.getTypeUtils() ).withSourceRHS( rightHandSide ),
+                                     null,
+                                     null,
+                                     false );
+            Assignment assignment = ctx
+                                   .getMappingResolver()
+                                   .getTargetAssignment(
+                                       method,
+                                       null,
+                                       targetType,
+                                       FormattingParameters.EMPTY,
+                                       criteria,
+                                       rightHandSide,
+                                       null,
+                                           () -> forgeMapping( rightHandSide, sourceType, targetType ) );
+            subClassMapping.setAssignment( assignment );
```

Comment:
construct and return the subClassMapping instead of creating it at the top and adding assignment later.